assets/src/design-system/components/switch/index.js

```diff
@@ -0,0 +1,229 @@
+  text-overflow: ellipsis;
+  white-space: nowrap;
+  transition: color 0.15s ease-out;
+
+  ${({ disabled }) =>
+    disabled &&
+    css`
+      cursor: default;
+      opacity: 0.3;
+    `}
+
+  /* add focus styling on the slider when the hidden input is focused */
+  :focus-within ~ span {
+    ${themeHelpers.focusCSS};
+  }
+`;
+
+const SlidingButton = styled.span`
+  position: absolute;
+  display: block;
+  top: 0;
+  left: 0;
+  width: 50%;
+  height: ${SWITCH_HEIGHT}px;
+  border-radius: 100px;
+  background-color: ${({ theme }) => theme.colors.interactiveBg.primaryNormal};
+  transition: left 0.15s ease-out;
+  z-index: 0;
+
+  ${({ hasOffset }) => hasOffset && `left: 50%`}
+`;
+
+/**
+ * A controlled radio group that looks like a switch.
+ *
+ * Keyboard navigation:
+ * - [Tab] can be used to navigate between radio groups
+ * - [Enter], [Space], [ArrowLeft], [ArrowRight] [ArrowUp], and [ArrowDown] can be used to select radio buttons
+ *
+ * @param {Object} props The props
+ * @param {string} props.className className that is applied to the outer container
+ * @param {string} props.disabled whether the radio buttons are disabled
+ * @param {string} props.groupLabel the label for the radio group. Used for accessibility
+ * @param {string} props.id unique id used for accessibility
+ * @param {string} props.name name of the radio group
+ * @param {string} props.offLabel label for the 'off' radio button
+ * @param {Function} props.onChange change event handler
+ * @param {string} props.onLabel label for the 'on' radio button
+ * @param {boolean} props.value the value of the radio group
+ * @return {Object} The radio button
+ */
+export const Switch = forwardRef(function (
+  {
+    className,
+    disabled,
+    groupLabel,
+    id,
+    offLabel,
+    onChange,
+    onLabel,
+    value,
+    ...props
+  },
+  ref
+) {
+  const radioGroupRef = useRef(ref);
+  const ids = useMemo(
+    () => ({
+      group: id || uuidv4(),
+      offInput: uuidv4(),
+      onInput: uuidv4(),
+    }),
+    [id]
+  );
+
+  const handleChange = useCallback(
+    (evt) => {
+      onChange(evt, evt.target.value === 'true');
+    },
+    [onChange]
+  );
+
+  const handleKeyDown = useCallback(
+    (evt) => {
+      onChange(evt, !value);
+    },
+    [onChange, value]
+  );
+
+  useKeyDownEffect(radioGroupRef, ['space', 'enter'], handleKeyDown, [
+    handleKeyDown,
+    value,
+  ]);
+
+  return (
+    <SwitchContainer
+      ref={radioGroupRef}
+      /* Class should contain "mousetrap" to enable keyboard shortcuts on inputs. */
+      className={'mousetrap ' + className}
+      role="radiogroup"
+      aria-labelledby={ids.group}
+    >
+      <VisuallyHiddenRadioGroupLabel id={ids.group}>
+        {groupLabel}
+      </VisuallyHiddenRadioGroupLabel>
+
+      <RadioButtonLabel
+        htmlFor={ids.onInput}
+        isActive={value}
+        disabled={disabled}
+      >
+        {onLabel}
+        <HiddenRadioButton
+          checked={value}
```

Comment:
this looks fine to me, but for some reason when I inspect the markup, the `checked` attribute isn't switching between the selected radio button:

https://user-images.githubusercontent.com/35983235/111491226-4e1a5e00-8701-11eb-934b-2add8b51987f.mp4

Reply:
Did some investigation and chatted with max about it. Here's our findings on this interesting looking bug:

There’s some really interesting talk about this online. It looks like it’s very common with the way people implement checkboxes:
https://github.com/facebook/react/issues/6321#issue-142960608
https://github.com/facebook/react/issues/3005#issuecomment-72513965

It's funny - really popular libraries do this too like Material UI
![material](https://user-images.githubusercontent.com/22185279/111506464-43ff5c00-870f-11eb-8abc-d7ac7d45f10b.gif)

---

So if we look at the values in the console, we see that checked is getting changed but it’s just not displayed in the tree
![radios](https://user-images.githubusercontent.com/22185279/111506616-698c6580-870f-11eb-9f5d-02ef8fcee468.gif)

This also happens to our other radio buttons (in design system and in the animation panel). We can fix this by adding a key that changes when the value is updated. This would invalidate the mounted component and re-mount a new one.

At the end of the day - in Javascript we use the `checked` property to determine the state of the checkbox. It is being updated correctly in javascript, but visually in the dom tree it is confusing 😕